doc: rough draft of RTR docs (#27284)
MaterializeInc/materialize: doc/user/content/get-started/isolation-level.md

It’s important to note that the linearizable guarantee only applies to transactions (including single statement SQL
queries which are implicitly single statement transactions), not to data written while ingesting from upstream sources.
So if some piece of data has been fully ingested from an upstream source, then it is not guaranteed to appear in the
next read transaction. See [real-time recency](#real-time-recency) and [strengthening correctness](https://github.com/MaterializeInc/materialize/issues/13107) for more details. If some
piece of data has been fully ingested from an upstream source AND is included in the results of some read transaction
THEN all subsequent read transactions are guaranteed to see that piece of data.

### Real-time recency

{{< private-preview />}}

> Materialize offers a form of "cross-system linearizability" known as real-time

Author: My apprehension is that "end-to-end" doesn't necessarily imply that it includes the source system. e.g. end-to-end _could_ be from ingestion to downstream propagation.

> Materialize offers a form of "cross-system linearizability" known as real-time

Reviewer: +1 on using "end-to-end". At least my impression from the ecosystem (_e.g._ Flink) is that this is exactly how you'd imply that any up- and downstream external systems are also considered.

> Materialize offers a form of "cross-system linearizability" known as real-time

Author: sounds good

recency. When using real-time recency, all client-issued `SELECT` statements
include at least all data visible to Materialize in any external source (i.e.,
sources created with `CREATE SOURCE` that use `CONNECTION`s, such as Kafka,

> MySQL, and PostgreSQL sources) shortly after Materialize receives the query.

Author: I will never find the right level of pedantry to please all people. :)

> MySQL, and PostgreSQL sources) shortly after Materialize receives the query.

Author: amended

> This is what we mean by linearizable––the results are guaranteed to contain all

Reviewer: ```suggestion
This is what we mean by linearizable — the results are guaranteed to contain all
```

Author: This is meant to be an "em dash," which is typically neither preceded nor followed by spaces.

visible data according to physical time.

For example, real-time recency ensures that if you have just performed an
`INSERT` into a PostgreSQL database that Materialize ingests as a source, all of
your real-time recency queries will include the just-written data in their

> results.

Author: This refers to a hypothetical insert into a PostgreSQL database consumed by Materialize, not Materialize itself.

Note, however, that we only guarantee the results will contain the data visible

> to Materialize. We do not guarantee that we will consider _only_ the data

Author: Will revise this to have fewer noise words.

> to Materialize. We do not guarantee that we will consider _only_ the data

Author: Amended

visible at the time we receive the query. For instance, the rate at which
Materialize ingests data might include additional data made visible after the

> timestamp we determined to be "real-time recent." Another example is that, due

Reviewer: ```suggestion
timestamp we determined to be "real-time recent". Another example is that, due
```

Author: Typically the period goes inside the quotation mark: https://www.hamilton.edu/academics/centers/writing/style/essentials/punctuation-of-quotations

> visible at the time we receive the query. For instance, the rate at which
> Materialize ingests data might include additional data made visible after the
> timestamp we determined to be "real-time recent." Another example is that, due
> to network latency, the timestamp from the external system that we determine to
> be "real-time recent" might be later than expected.

Reviewer 2: ```suggestion
visible at the time we receive the query. The "real-time recent" results may 
include additional data made visible to Materialize before and even soon 
after the timestamp we determined to be "real-time recent." 
This is due to the distributed, streaming nature of the system: More data may 
be made available to Materialize in parallel to the query being prepared and 
submitted, or while Materialize makes a request to the external system to 
get the timestamp to use; Once the timestamp is obtained, Materialize may 
already have read data to a later timestamp. 
All in all, the guarantee is only to consider _at least_ the data available at 
the time the query was submitted.
```

Author: I left this one alone––feel free to rewrite as you see fit once this merges.

Because Materialize waits until it ingests the data from the external system,
real-time recency queries can have additional latency. This latency is
introduced by both the time it takes us to ingest and commit data from the

> source and the time spent communicating with it to determine what data it has

Reviewer: ```suggestion
source, and the time spent communicating with it to determine what data it has
```

Author: This would be an oxford comma, which is not commonly used with lists of two items.

made available to us (e.g., querying PostgreSQL for the replication slot's LSN).

**Details**

> - Enable this feature per session using `SET REAL_TIME_RECENCY TO TRUE`.

Reviewer: Edited the edit, since we standardize on `=` in the documentation.

- Control the timeout for connecting to the external source to determine the

>   timestamp with the `REAL_TIME_RECENCY_TIMEOUT` session variable.

Reviewer: This also needs to be added to the [`SET`](https://materialize.com/docs/sql/set/#other-configuration-parameters) command page. I'll push a commit shortly.

- Real-time recency queries only guarantee visibility of data from external
  systems (e.g., sources like Kafka, MySQL, and PostgreSQL). Real-time recency
  queries do not offer any form of guarantee when querying Materialize-local

>   objects, such as LOAD GENERATOR sources or system tables.

Reviewer: ```suggestion
  objects, such as [`LOAD GENERATOR`sources](/sql/create-source/load-generator/) or system tables.
```

- Each real-time recency query connects to each external source transitively
  referenced in the query. The more external sources that are referenced, the

>   greater the likelihood of network-induced latency.

Reviewer 2: ```suggestion
  greater the likelihood of additional latency, both network-induced and due to 
  needing to wait for more data to be ingested.
```

Author: Amended

> - Each real-time recency query connects to each external source transitively
>   referenced in the query. The more external sources that are referenced, the
>   greater the likelihood of network-induced latency.

Reviewer: ```suggestion
- Each real-time recency query connects to each external source transitively
  referenced. This means that the latency overhead will be proportional to the
  number of sources referenced in the query.
```

Author: These queries occur in parallel (or at least concurrently) so the latency overhead is not proportional to the number of items queried, but the latency _is_ the max of all of the requests. Not sure if there's a better way to make that clear.

- Materialize doesn't currently offer a mechanism to provide a "lower bound" on
  the data we consider to be "real-time recent" in an external source. Real-time
  recency queries return at least all data visible to Materialize when our

>   client connection communicates with the external system.

Author: Ah, sorry––this is in the strict serializable session, which I realize now was too subtle a clue that this was only available at that isolation level. Amended.
